### DISCUSSION TITLE
Is `thread_sensitive` required for ORM calls in async consumers?

### DISCUSSION THREAD

**Discussion by Archmonger** [django/channels#2051] [Q&A] [answered]
Andrew Godwin seems [to have implied](https://github.com/django/asgiref/issues/218#issuecomment-730505886) that the Django ORM may not need `thread_sensitive=True` within async consumers.

> **[andrewgodwin](https://github.com/andrewgodwin) commented [on Nov 19, 2020](https://github.com/django/asgiref/issues/218#issuecomment-730505886)**
> I think Channels would mostly be safe as there's nothing synchronous on the main thread to lock to, but I'd have to go back into the codebase to be sure, and it's been a loooong time since I actively developed in there.

We are trying to figure out if we can safely turn off `thread_sensitive` on ReactPy's [`use_query`](https://reactive-python.github.io/reactpy-django/3.5.1/reference/hooks/#use-query) and [`use_mutation`](https://reactive-python.github.io/reactpy-django/3.5.1/reference/hooks/#use-mutation) hooks.

**Comment by carltongibson** [comment, marked as answer]
I take that as referring only to the consumer related code. Django's ORM definitely requires use of `thread_sensitive=True`, but the effect of this is mitigated since Django introduced per-request async contexts (so it's one thread per-request, rather than being single threaded.)

**Reply by Archmonger** [reply]
Are the per-request contexts functional within Channels? My interpretation of that `django/django` change seems to only have enabled that within the `django/django` stack.

More specifically, since `channels` operates outside of the traditional Django stack shouldn't that mean per-request contexts don't exist?

**Reply by carltongibson** [reply]
Once Channels passes down to `ASGIHandler` it **is** using the Django stack. 

Otherwise in consumer code, you can decide whether you need `thread_sensitive` when using `sync_to_async()` (but if you're using the ORM, you do.)
